adagrams/game.py

```diff
@@ -1,11 +1,120 @@
+    for i in range(number_of_letters):
+        random_index = random.randint(0, len(all_letter_list) - 1)
+        player_letters_list[i]= all_letter_list.pop(random_index)
+    return player_letters_list
 
 def uses_available_letters(word, letter_bank):
-    pass
+    '''
+    This function has two parameters: input word (string) letter_bank (array of drawn letters from draw_letters function)
+    Returns True if every letter in the input word is in the letter_bank
+    Returns False if there is a letter in input that is not present in the letter_bank or has too much of compared to the letter_bank
+    '''
+    checked_word = list()
+    word_ls = []
+    word_ls[:] = word.upper()
+    for letter in word_ls:
+        if letter in letter_bank:
+            checked_word.append(letter)
+            index = letter_bank.index(letter)
+            del letter_bank[index]
+            if checked_word == word_ls:
+                letter_bank.extend(checked_word)
+                letter_bank.sort()
+                return True
+    return False
 
 def score_word(word):
-    pass
+    '''
+    This function has one parameter: word, which is a string of characters
+    Returns an integer representing the number of points
+    Each letter within word has a point value (located in SCORE_CHART dictionary)
+    If the length of the word is 7, 8, 9, or 10, then the word gets an additional 8 points
+    '''
+    user_points = 0
+    word_ls = []
+    word_ls[:] = word.upper()
+    if len(word_ls)>= 7:
+        user_points = 8
+    for element in SCORE_CHART:
+        for letter in word_ls:
+            if letter in SCORE_CHART[element]:
+                user_points += element
+    return user_points
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    '''
+    This function has one parameter - word_list( list of strings)
+    Returns a tuple that represents the data of a winning word and it's score.
+    In the case of tie in scores:
+        1 - the word which has 10 letters
+        2 - prefer the word with the fewest letters
+        3 - the first one in the supplied list
+    '''
+    users_score = 0
+    max_score = 0
+    max_word = ''
+    for word in word_list:
+        users_score = score_word(word)
+        if users_score > max_score:
+            max_score = users_score
+            max_word = word
+        elif users_score == max_score:
+            if len(word) < len(max_word):
+                if len(max_word) == 10:
+                    continue
+                else:
+                    max_score = users_score
+                    max_word = word
+            elif len(word) == len(max_word):
+                word_index = word_list.index(word)
+                max_word_index = word_list.index(max_word)
+                if word_index > max_word_index:
+                    continue
+                else:
+                    max_word = word
+                    max_score = users_score
```

Comment:
A for-in loop will iterate through a list from index 0 until the end of the list (in order), which means that `word_index` will always be greater than `max_word_index`. If we know that 2 words have the same score and the same length, do we ever need to update variables in lines 109-115? What happens if we remove that code?